rx/concurrency/newthreadscheduler.py

```diff
@@ -98,21 +98,23 @@ def schedule_periodic(self,
             recurring action (best effort).
         """
 
-        secs: float = self.to_seconds(period)
+        seconds: float = self.to_seconds(period)
+        timeout: float = seconds
         disposed: threading.Event = threading.Event()
 
-        s = state
-
         def run() -> None:
+            nonlocal state, timeout
             while True:
-                disposed.wait(secs)
+                if timeout > 0.0:
+                    disposed.wait(timeout)
                 if disposed.is_set():
                     return
 
-                nonlocal s
-                new_state = action(s)
-                if new_state is not None:
-                    s = new_state
+                time: typing.AbsoluteOrRelativeTime = self.now
+
+                state = action(state)
+
+                timeout = seconds - (self.now - time).total_seconds()
```

Comment:
If I understand correctly, we are working with interval defined as the time elapsed between consecutive action starts, not the time elapsed between the end of the previous action and the start of the new one. Don't you think it can be dangerous if the execution time of an action is greater than the interval time?

Reply:
Well, I just thought that users (or me at least) would expect the interval to include the execution time. I tried to make sure that potentially negative values are clipped to zero. Not in the subtraction here, though, but at the start of subsequent schedule calls — that way regular calls are clipped as well.

Does that make sense?

Alignment with the blog is entirely coincidental, but I’ll take a bonus :-)

Reply:
Yes it does! Thanks.